### ISSUE TITLE
Too many opened file handles to spotbugs plugins

### ISSUE THREAD

**Issue by svilenvul** [spotbugs/spotbugs-maven-plugin#35]
We are running the spotbugs-maven-plugin programmatically in a multi-module Maven project. This is why it is difficult to describe the exact spotbugs maven plugin configuration, this is why I will attach part of the Maven debug logs, where this can be seen:

[spotbugs-config.txt](https://github.com/spotbugs/spotbugs-maven-plugin/files/1796287/spotbugs-config.txt)

While trying to set up a Jenkins job, we faced ‘Too many open files’  problem causing the build to fail.

We started investigating the problem and found out that there is a huge number of opened files (spotbugs plugins).

I attach here the output of the following command `ls -l /proc/MAVEN_PID/fd/ > openfiles.txt`:

[openfiles.txt](https://github.com/spotbugs/spotbugs-maven-plugin/files/1796374/openfiles.txt)

They are all located in the target directory.

Used spotbugs-maven-plugin/3.1.3 with spotbugs 3.1.2

**Comment by KengoTODA** [comment]
note from my investigation:
* `ICodeBase`,`IClassPath` is `AutoCloseable`.
* `IClassPath` is responsible to close all `ICodeBase` instances inside it
* `FindBugs2#clearCaches()` closes generated `IClassPath`
* It seems that this plugin doesn't invoke `FindBugs2#clearCaches()` for now

So I guess that we should invoke this `FindBugs2#clearCaches()` method, or create classloader to GC `FindBugs2.classpath` that is singleton field?

**Comment by KengoTODA** [comment]
https://github.com/spotbugs/spotbugs/issues/589 is similar with this issue, but this issue's target is mainly jar files in classpath to scan.

**Comment by KengoTODA** [comment]
@svilenvul let me confirm: are your Maven uses `-T` option, to run modules in parallel?

**Comment by KengoTODA** [comment]
WIP: https://github.com/spotbugs/spotbugs/compare/fix-resource-leaks

**Comment by svilenvul** [comment]
> @svilenvul let me confirm: are your Maven uses -T option, to run modules in parallel?

No, currently it is not being used.

**Comment by KengoTODA** [comment]
> We started investigating the problem and found out that there is a huge number of opened files (spotbugs plugins).
Almost all opened file isn't not spotbugs plugin, but libraries in aux classpath.

In current architecture, spotbugs keeps all jar files opened during analysis (because `IClassPath` is keep opened), so we should face problem like this. To fix this issue, I think we need to change internal architecture to handle libraries in aux & app classpaths as stream.

**Comment by famod** [comment]
Has there been any progress? We are now seeing this frequently on our Jenkins CI server (we are using -T and there can also be concurrent builds).

**Comment by hazendaz** [comment]
This issue is likely related to #535 that has existed in this code base since its original inception with findbugs.  Has to do with the -T and it getting locks as system.in was being called.  This is to be removed in upcoming release 4.7.3.1 within the week.

**Comment by hazendaz** [comment]
closing as believe fixed with 4.7.3.1.  If still an issue, please open a new ticket based off current data.
